Intro R revamp (PR 4 of n) : Add third notebook (#168)
AlexsLemonade/training-modules: intro-to-R-tidyverse/03-intro_to_tidyverse.Rmd

---

> title: "Intro to ggplot2"

Author: Whoops. I feel like I do this 99% of the time.

that we performed a set of [differential expression analyses on](./scripts/00-setup-intro-to-R.R).  

**More tidyverse resources:**  
- [R for Data Science](https://r4ds.had.co.nz/)  
- [Tidyverse documentation](https://dplyr.tidyverse.org/)  
- [Cheatsheet of Tidyverse data transformation](https://github.com/rstudio/cheatsheets/raw/master/data-transformation.pdf)  
- [Online Tidyverse book chapter](https://privefl.github.io/advr38book/tidyverse.html)  

## Set Up

The `tidyverse` is a collection of packages that are handy for general data 
wrangling, analysis, and visualization. 
Other packages that are specifically handy for different biological analyses are 
found on [Bioconductor](https://www.bioconductor.org/).
If we want to use a package's functions we first need to install them.

TODO: Language around Docker container may change depending on RStudio cloud stuff

In our Docker container, we already have `tidyverse` and other packages we will
use in this workshop installed for you. 
But if you needed to install it or other packages available on CRAN, you 

> do it using the `install.packages` function like this: `install.packages("tidyverse")`

Author: Makes sense!

created only *if* `dir.exists(results)` is NOT equal to `TRUE`.

```{r}
# If 'results' directory doesn't exist...
if (!dir.exists("results")) {
  # ... create a 'results' directory
  dir.create("results")
}
```

Note that the `dir.exists` function will not work on files themselves.
In this case, there is an analogous function called `file.exists`.

Try using the `file.exists` function to see if the file `gene_results_GSE44971.tsv` file 
exists in the current directory.
Use the code chunk we set up for you below. 
Note that in our notebooks (and sometimes elsewhere), wherever you see a 
`<FILL_IN_THE_BLANK>` like in the chunk below, that is meant for you to replace 
with the correct phrase before you run the chunk (otherwise you will get an error).

```{r eval=FALSE}
# Replace the <PUT_FILE_NAME_HERE> with the name of the file you are looking for
# Remember to use quotes to make it a character string
file.exists(<PUT_FILE_NAME_HERE>)
```

Now that we've determined that `gene_results_GSE44971.tsv` exists, we are ready to read it 
into our R environment.

#### Read a TSV file

Declare the name of the directory where we will read in the data. 

```{r}
data_dir <- "data"
```

> In R, we will use `file.path()` for this (more on this later in this notebook).  

Author: Ah this is a remnant that I should have deleted.

Although base R has read-in functions, the `readr` functions are faster and more
straightforward to use so we are going to use those here. 
Because the file we are reading in is a TSV (tab separated values) file we will 
be using the `read_tsv` function. 
But note that there are analogous functions for CSV (comma separated values) 
files (`read_csv`) and other files types.

## Read in the differential expression analysis results file

Here we are using a `tidyverse` function `read_tsv` from a library of functions called `readr`. 

```{r}
stats_df <- readr::read_tsv(file.path(data_dir,
                                      "gene_results_GSE44971.tsv"))
```

> ```{r}

Author: I like it.

```{r}
filter(stats_df, contrast == "male_female")
```  

...is the same as the output from this:  

```{r}
stats_df %>% filter(contrast == "male_female")
```  
  
This makes it so your code is cleaner and easier to read. 
Let's look at an example with our stats of of how the same 
functions look with or without pipes:

*Example 1:* without pipes: 

```{r}
stats_nopipe <- stats_df
stats_nopipe <- arrange(stats_nopipe, t_statistic)
stats_nopipe <- filter(stats_nopipe, avg_expression > 5)
stats_nopipe <- select(stats_nopipe, contrast, log_fold_change, p_value)
```
  
UGH, we have to repeat and assign `stats_nopipe` so many times here! 
It's annoying and makes it harder for people to read. 
  
*Example 2:* Same result as 1 but with pipes!

```{r}
# Example of the same modifications as above but with pipes!
stats_pipe  <- stats_df %>%
               arrange(t_statistic) %>%
               filter(avg_expression > 5) %>%
               select(contrast, log_fold_change, p_value)
```

Let's double check that these are the same by using the function, all.equal: 

```{r}
all.equal(stats_nopipe, stats_pipe)
```

`all.equal` is letting us know that these two objects are the same. 

> (You can use all.equal in other instances and adjust the parameters depending on

Author: I'll delete it. Seems unnecessary at this point.

data.frame it originated from.

## Save data to files

#### Save to TSV files

Let's write data we edited to a file.
To do this, we can use the `readr` library of `write` functions. 
Note that the second argument of `write_tsv` needs to be a character string that 
contains our `file.path` to the new file we would like to create.
Remember that we created a `results` directory to put our output in. 
But if we want to save our data to a directory other than our working directory, 
we need to specify this. 
This is what we will use the `file.path` function for. 
Let's look at what `file.path` does, by taking a look at the print out of it 
from examples below.

```{r}
# Which of these file paths is what we want to use to save our data to the
# results directory we created at the beginning of this notebook?
file.path("docker-install", "stats_summary.tsv")
file.path("results", "stats_summary.tsv")
file.path("stats_summary.tsv", "results")
```

Replace `<NEW_FILE_PATH>` below with the `file.path` from above that will 
successfully save our file to the `results` folder 

```{r eval=FALSE}
# Write our data.frame to a TSV file

> write_tsv(stats_summary_df, <NEW_FILE_PATH>)

Reviewer: ```suggestion
readr::write_tsv(stats_summary_df, <NEW_FILE_PATH>)
```

```

Check in your `results` directory to see if your new file has successfully saved.

#### Save to RDS files

> For this example we have been working with `data.frame`s, however, in other 
> situations we may want to save more complicated structure objects or very large 

Reviewer: ```suggestion
For this example we have been working with data frames, which are conveniently 
respresented as TSV or CSV tables. 
However, in other situations we may want to save more complicated or very large 
```

objects. 
RDS (R Data Serialized/Single) files may be a better option in these instances.
RDS is R's special file format for holding data exactly as you have it in your 
R environment. 
RDS files can also be compressed, meaning they will take up less space on your 
computer. 
Let's save our data to an RDS file in our `results` folder.
You will need to replace the `.tsv` with `.RDS`, but you can use what we 

> determined as our `file.path` for the last chunk as your template. 

Reviewer: ```suggestion
determined as our file path for the last chunk as your template. 
```

```{r eval=FALSE}
# Write your object to an RDS file

> write_rds(stats_summary, <PUT_CORRECT_FILE_PATH_HERE>)

Reviewer: ```suggestion
readr::write_rds(stats_summary, <PUT_CORRECT_FILE_PATH_HERE>)
```

## A brief intro to the `apply` family of functions

In base R there the `apply` family of functions can be an alternative 
methods for performing transformations across a data frame, matrix or for other 
object structures. 

One of this family is (shockingly) the function `apply()`, which operates on 
matrices or data frames. 
The first argument to `apply()` is the data object we want to work on.
The third argument is the function we will apply to each row or column of the 
data object.
The second argument in specifies whether we are applying the function 
across rows or across columns (1 for rows, 2 for columns).

Remember that `gene_df` is a gene x sample gene expression data frame. 

```{r}
# Calculate row means
gene_means <- apply(gene_df[, -1], 1, mean) # Notice we are using 1 here

# How long will `gene_means` be? 
length(gene_means)
```

Why do we have a `[, -1]` after `gene_df` in the above chunk?

Now let's investigate the same set up, but use 2 to `apply` over the columns of
our matrix.

```{r}
# Calculate sample means
sample_means <- apply(gene_df[, -1], 2, mean) # Notice we use 2 here

# How long will `sample_means` be? 
length(sample_means)
```

> Although the `apply` functions may not be as easy to use as the `tidyverse` 

Reviewer: I don't know whether we want to style "tidyverse" as `tidyverse`, but it seems like we were not doing that much, and we should try to be consistent. Perhaps we should refer to "the Tidyverse" in text, and `tidyverse` only when specifically refering to the package.
```suggestion
Although the `apply` functions may not be as easy to use as the tidyverse 
```

functions, for some applications, `apply` methods can be better suited.
In this workshop, we will not delve too deeply into the various other apply 
functions (`tapply()`, `lapply()`, etc.) but you can read more examples about 
them [here](https://www.guru99.com/r-apply-sapply-tapply.html).

## The dplyr::join functions

Let's say we have a scenario where we have two data frames that we would like to 
combine. 
Recall that `stats_df` and `gene_df` are data frames that contain information 
about some of the same genes.
The [`dplyr::join` family of functions](https://dplyr.tidyverse.org/reference/join.html) 
are useful for various scenarios of combining data frames. 

For now, we will focus on `inner_join()`, which will combine data frames by only

> keeping information about the genes that are in both data frames. 

Reviewer: ```suggestion
keeping information about matching rows that are in both data frames. 
```

We need to use the `by` argument to designate what column(s) 
should be used as a key to match the data frames.

> In this case we want to match values in the `ensembl_id` column from `stats_df` 

Reviewer: ```suggestion
In this case we want to match the gene information between the two, so we will specify 
that we want to compare values in the `ensembl_id` column from `stats_df` 
```

to the `Gene` column from `gene_df`.

```{r}
stats_df %>% 
  inner_join(gene_df, by = c('ensembl_id' = 'Gene')) 
```

## Save data to files

#### Save to TSV files

> Let's write data we edited to a file.

Reviewer: ```suggestion
Let's write some of the data frames we created to a file.
```
